### PR TITLE
Add HTCondor batch system

### PR DESCRIPTION
**Pull request by jasoncpatton**
Current version works with HTCondor pools that use a shared file system (this includes, for example, a personal HTCondor pool on a single machine). Future versions will work with distributed HTCondor pools.

Requires the HTCondor Python bindings, which are either installed by the system package manager as part of a HTCondor install or can be installed separately via `pip install htcondor`.

Passes tests in `src/toil/test/batchSystems/batchSystemTest.py` when tested on a machine with a personal HTCondor pool.

To do to make this plugin more complete (will open issue): 
1. Transfer local file stores to the execute nodes.
2. Check pool resources.
3. Support running in the Open Science Grid.

### CODE REVIEW

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by benedictpaten** [comment]
This is great to see. @jasoncpatton Can you take a look at the failures Jenkins is reporting? If there are failures you don't understand we can try to find some developer time to help out to get this merged!

**Comment by jasoncpatton** [comment]
I see what the problem is, but I'm not sure how to address it.

The `import htcondor` statement won't (and shouldn't) work on the test system since htcondor is not a required package for Toil as a whole, though it is required for using the HTCondor batch system plugin. How can this be handled gracefully? Should htcondor be another install option (like mesos)?

**Comment by mr-c** [comment]
@jasoncpatton Yes, it should be an additional install option https://github.com/BD2KGenomics/toil/blob/master/setup.py#L19

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by jasoncpatton** [comment]
@ejacox Should I add htcondor to the extras in jenkins.sh? (Or is that "cheating"?)

**Comment by ejacox** [comment]
@jasoncpatton, That's exactly right. Add it to the "make develop ..." line so that Jenkins will install it before running a test.

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by jasoncpatton** [comment]
That should cover all the types of exceptions that our python bindings can raise when they can't find the condor scheduler process or config.

**Comment by ejacox** [comment]
Jenkins, test this please.

**Comment by ejacox** [comment]
Jenkins, add to whitelist

**Comment by jasoncpatton** [comment]
I could use some pointers in debugging these tests, it's not clear to me how (if) they are related to the new batch system (and batch system test) code.

**Comment by ejacox** [comment]
We have been getting more transient errors lately in our AWS tests. We will track down the problems when we have time. In the mean time, we rerun the tests or ignore them.

**Comment by ejacox** [comment]
Jenkins, test this please.

**Comment by ejacox** [comment]
@jasoncpatton Is it ready to merge? We can ignore the AWS failures.

File: src/toil/batchSystems/htcondor.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        localID = self.handleLocalJob(jobNode)
+        if localID:
+            return localID
+        else:
+            self.checkResourceRequest(jobNode.memory, jobNode.cores, jobNode.disk)
+            jobID = self.getNextJobID()
+            self.currentJobs.add(jobID)
+
+            # Add the jobNode.disk and jobNode.jobName to the job tuple
+            self.newJobsQueue.put((jobID, jobNode.cores, jobNode.memory, jobNode.disk, jobNode.jobName, jobNode.command))
```
Comment (mr-c) [inline src/toil/batchSystems/htcondor.py]: General question for Toil devs, why don't we put the `jobNode` itself in the `newJobsQueue`?

**Comment by jasoncpatton** [comment]
Yes

File: src/toil/batchSystems/htcondor.py (hunk 1 above)
Comment (ejacox) [reply src/toil/batchSystems/htcondor.py]: Toil only needs the jobId. It seems the queue was only jobIds originally and later changed to a tuple to include extra information used by the batch system.

**Comment by ejacox** [comment]
Thank you @jasoncpatton

**Pull request merged**